ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables.hs

```diff
@@ -0,0 +1,388 @@
+  projectLedgerTables _ = trivialLedgerTables
+
+  -- | Overwrite the tables in the given ledger state.
+  --
+  -- The contents of the tables should not be /younger/ than the content of the
+  -- ledger state. In particular, for a
+  -- 'Ouroboros.Consensus.HardFork.Combinator.Basics.HardForkBlock' ledger, the
+  -- tables argument should not contain any data from eras that succeed the
+  -- current era of the ledger state argument.
+  --
+  -- The constraints on @mk@ are necessary because the 'CardanoBlock' instance
+  -- uses them.
+  withLedgerTables ::
+       (CanMapMK mk, CanMapKeysMK mk, ZeroableMK mk)
+    => l any
+    -> LedgerTables l mk
+    -> l mk
+  default withLedgerTables ::
+       LedgerTablesAreTrivial l
+    => l any
+    -> LedgerTables l mk
+    -> l mk
+  withLedgerTables st _ = convertMapKind st
+
+instance ( Ord (Key l)
+         , Eq (Value l)
+         , Show (Key l)
+         , Show (Value l)
+         , NoThunks (Key l)
+         , NoThunks (Value l)
+         ) => HasLedgerTables (LedgerTables l) where
+  projectLedgerTables = castLedgerTables
+  withLedgerTables _ = castLedgerTables
+
+-- | Convenience class, useful for partially applying the composition of
+-- 'HasLedgerTables' and 'Ticked1'.
+type HasTickedLedgerTables :: LedgerStateKind -> Constraint
+class HasLedgerTables (Ticked1 l) => HasTickedLedgerTables l where
+instance HasLedgerTables (Ticked1 l) => HasTickedLedgerTables l
+
+-- | LedgerTables are projections of data from a LedgerState and as such they
+-- can be injected back into a LedgerState. This is necessary because the Ledger
+-- rules are unaware of UTxO-HD changes. Thus, by stowing the ledger tables, we are
+-- able to provide a Ledger State with a restricted UTxO set that is enough to
+-- execute the Ledger rules.
+--
+-- In particular, HardForkBlock LedgerStates are never given diretly to the
+-- ledger but rather unwrapped and then it is the inner ledger state the one we
+-- give to the ledger. This means that all the single era blocks must be an
+-- instance of this class, but HardForkBlocks might avoid doing so.
+type CanStowLedgerTables :: LedgerStateKind -> Constraint
+class CanStowLedgerTables l where
+
+  stowLedgerTables     :: l ValuesMK -> l EmptyMK
+  default stowLedgerTables ::
+       (LedgerTablesAreTrivial l)
+    => l ValuesMK
+    -> l EmptyMK
+  stowLedgerTables = convertMapKind
+
+  unstowLedgerTables   :: l EmptyMK  -> l ValuesMK
+  default unstowLedgerTables ::
+       (LedgerTablesAreTrivial l)
+    => l EmptyMK
+    -> l ValuesMK
+  unstowLedgerTables = convertMapKind
+
+{-------------------------------------------------------------------------------
+  Serialization Codecs
+-------------------------------------------------------------------------------}
+
+-- | This class provides a 'CodecMK' that can be used to encode/decode keys and
+-- values on @'LedgerTables' l mk@
+--
+-- TODO: can this be removed in favour of EncodeDisk and DecodeDisk?
```

Comment:
> After some thought: when are we ever serializing non-EmptyMK ledger tables? Maybe we don't need any codecs for whole ledger tables? (As opposed to the individual key/value codecs for the backing store.)

I'm not sure, but I *think* we should be able to replace uses of `CodecMK` by requiring encoding/decoding constraints on `Key l` and `Value l`. 

Historically, we did not have `Key l` and `Value l` and the types of UTXO keys/values were "hidden" to the consensus layer. In that context it made more sense to have `CodecMK`, because there were no type families `Key`/`Value` to require constraints for

Reply:
> After some thought: when are we ever serializing non-EmptyMK ledger tables? Maybe we don't need any codecs for whole ledger tables?

There are no codes for whole ledger tables. This is a LedgerTable of codecs, not a codec of LedgerTables.

Reply:
https://github.com/IntersectMBO/ouroboros-consensus/issues/1309